File: include/restricted-extras-qscintilla/libqscistyledtext.zig
const C = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qscistyledtext.html
pub const qscistyledtext = struct {
    /// New constructs a new QsciStyledText object.
    ///
    /// ``` text: []const u8, style: i32 ```
    pub fn New(text: []const u8, style: i32) ?*C.QsciStyledText {
        const text_str = C.struct_libqt_string{
            .len = text.len,
            .data = @constCast(text.ptr),
        };

        return C.QsciStyledText_new(text_str, @intCast(style));
    }

    /// New2 constructs a new QsciStyledText object.
    ///
    /// ``` text: []const u8, style: ?*C.QsciStyle ```
    pub fn New2(text: []const u8, style: ?*anyopaque) ?*C.QsciStyledText {
        const text_str = C.struct_libqt_string{
            .len = text.len,
            .data = @constCast(text.ptr),
        };

        return C.QsciStyledText_new2(text_str, @ptrCast(style));
    }

    /// New3 constructs a new QsciStyledText object.
    ///
    /// ``` param1: ?*C.QsciStyledText ```
    pub fn New3(param1: ?*anyopaque) ?*C.QsciStyledText {
        return C.QsciStyledText_new3(@ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qscistyledtext.html#apply)
    ///
    /// ``` self: ?*C.QsciStyledText, sci: ?*C.QsciScintillaBase ```
    pub fn Apply(self: ?*anyopaque, sci: ?*anyopaque) void {
        C.QsciStyledText_Apply(@ptrCast(self), @ptrCast(sci));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qscistyledtext.html#text)
    ///
    /// ``` self: ?*C.QsciStyledText, allocator: std.mem.Allocator ```
    pub fn Text(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QsciStyledText_Text(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qscistyledtext.html#style)
    ///
    /// ``` self: ?*C.QsciStyledText ```
    pub fn Style(self: ?*anyopaque) i32 {
        return C.QsciStyledText_Style(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QsciStyledText ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QsciStyledText_Delete(@ptrCast(self));
    }
};

File: src_original/restricted-extras-qscintilla/libqscistyledtext.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qscistyledtext.h>
#include "libqscistyledtext.h"
#include "libqscistyledtext.hxx"

QsciStyledText* QsciStyledText_new(const libqt_string text, int style) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    return new QsciStyledText(text_QString, static_cast<int>(style));
}

QsciStyledText* QsciStyledText_new2(const libqt_string text, const QsciStyle* style) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    return new QsciStyledText(text_QString, *style);
}

QsciStyledText* QsciStyledText_new3(const QsciStyledText* param1) {
    return new QsciStyledText(*param1);
}

void QsciStyledText_Apply(const QsciStyledText* self, QsciScintillaBase* sci) {
    self->apply(sci);
}

libqt_string QsciStyledText_Text(const QsciStyledText* self) {
    const QString _ret = self->text();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

int QsciStyledText_Style(const QsciStyledText* self) {
    return self->style();
}

void QsciStyledText_Delete(QsciStyledText* self) {
    delete self;
}
